### PR TITLE
Service Validations - Adds error message for FormError fields

### PR DESCRIPTION
**Pull request by dthyresson**
Fixes https://github.com/redwoodjs/redwood/issues/4371

WIP.

Need to document any changes (the "name" is the field input name) and now message substitution keys are humanized and titlecased).

```ts
  validate(input.email, 'email', {
    email: true,
  })
```

Here the second argument is the form input name (case-sensitive) to match the key for the form-input level message. This is also really the Prisma input argument name in same case when used in the service.

Then, that value will be humanized and titlecased to be "Email must be formatted like an email address"

If the field was `workEmail` or `work_email` and you also had `email`, then:

```ts
  validate(input.email, 'workEmail', {
    email: true,
  })
```

`"Work Email must be formatted like an email address"`

### CODE REVIEW

**Comment by dthyresson** [comment]
@cannikin This is ready for review.

I am not sure about 

TODO:

- [ ] test for the field level message in GraphQL response? Add to e2e?
- [ ] test any changes for RedwoodRecord


The RedwoodRecord tests pass, but I have been having problems using RedwoodRecord in an app to test out.

@dac09 I also think maybe the e2e test needs to be updated perhaps to check for service api-side error messages?

And last @thedavidprice a slight change to the docs would need to be done as:

```ts
validate(input.email, { email: { message: 'Please provide a valid email address' } })

// Three argument form: validate(value, name, validations)
validate(input.email, 'Email Address', { email: true }
```

The `Email Address` here would now be `email` to make the input name and if you wanted "Email Address" you'd have to:

```ts
validate(input.email, 'email', { email: { 'Email Address is invalid' }
```

I *must* have the exact key/input name to match the message with the form input field. The first argument is the value, but the field.

File: packages/api/src/validations/validations.ts [hunk 1]
```diff
@@ -571,7 +571,7 @@ export const validateWith = (func: () => void) => {
     func()
   } catch (e) {
     const message = (e as Error).message || (e as string)
-    throw new ValidationErrors.ServiceValidationError(message)
+    throw new ValidationErrors.ServiceValidationError(message, message)
```
Comment (cannikin) [inline packages/api/src/validations/validations.ts:574]: Hmm, isn't the second arg to `ServiceValidationError` an object for string substitutions? How does this work if you just pass the `message` string twice?

**Comment by dthyresson** [comment]
> It'd be nice to see a test or two with `camelCase` or `snake_case` field names to see that they're humanized and titlecased as expected!

Good point. Will add. Thanks.

File: packages/api/src/validations/validations.ts (hunk 1 above)
Comment (dthyresson) [reply packages/api/src/validations/validations.ts:574]: Oh that may be a mistake -- I originally had code that passed two messages: a main and field one, but that got too complicated for this round to manage (and would change the signature of the validate, so I abandoned).

**Comment by dthyresson** [comment]
> > It'd be nice to see a test or two with `camelCase` or `snake_case` field names to see that they're humanized and titlecased as expected!
> 
> Good point. Will add. Thanks.

Tests added!

**Review by cannikin — APPROVED** [review]

**Comment by cannikin** [comment]
Awesome! Now to update the tutorial to actually use these!

**Comment by thedavidprice** [comment]
MERGED

**Pull request merged**